awsutil/options.go

```diff
@@ -0,0 +1,161 @@
+package awsutil
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/hashicorp/go-hclog"
+)
+
+// getOpts iterates the inbound Options and returns a struct
+func getOpts(opt ...Option) (options, error) {
+	opts := getDefaultOptions()
+	for _, o := range opt {
+		if o == nil {
+			continue
+		}
+		if err := o(&opts); err != nil {
+			return options{}, err
+		}
+	}
+	return opts, nil
+}
+
+// Option - how Options are passed as arguments
+type Option func(*options) error
+
+// options = how options are represented
+type options struct {
+	withEnvironmentCredentials bool
+	withSharedCredentials      bool
+	withAwsSession             *session.Session
+	withClientType             string
+	withUsername               string
+	withAccessKey              string
+	withSecretKey              string
+	withLogger                 hclog.Logger
+	withStsEndpoint            string
+	withIamEndpoint            string
+	withMaxRetries             *int
+	withRegion                 string
+	withHttpClient             *http.Client
+}
+
+func getDefaultOptions() options {
+	return options{
+		withEnvironmentCredentials: true,
```

Comment:
Do functional options tend to strictly stick with `With*`, or can we swap these to something like `SkipEnvironmentCredentials` so that the default value is also the zero value (false)?

Something like this seems to be a bit better to read, especially if this is for opt-out logic.
```
SkipEnvironmentCredentials(true)
```

Reply:
Default value isn't related to `With`. The function takes in a boolean, so it's not `WithEnvironmentCredentials` always results in true, it results in whatever you pass it.

Reply:
Right, I realize that `getDefaultOptions` will set these to `true` if that's not provided and will honor what's being provided otherwise. I just thought that it'd be nicer to have something like this

```
SkipEnvironmentCredentials(true)
or
SkipEnvironmentCredentials()
```

than

```
WithEnvironmentCredentials(false)
```
since this behaves as an opt-out. This is all subjective though, so I'm fine with leaving as-is.

Reply:
I got convinced by some other people that functions toggling booleans should always take in boolean parameters instead of no parameters, since you may want to explicitly toggle something to the same as the default (e.g. for clarity, or if you don't know what the default is). Hence sticking with `With`. As you said it's subjective, so I'm going to leave it this way, but I get where you're coming from :-)